### PR TITLE
Added apoc.merge.node.eager() and apoc.merge.relationship.eager()

### LINKED ISSUE #1182: apoc.merge.node() does not update node properties when matching to existing node
## How to Reproduce the Problem
On an empty database using APOC 3.5.0.3 execute the following:

CREATE (p:Person{ssid:'123', name:'Jim'}) 
CALL apoc.merge.node(['Person'],{ssid:'123'}, {name:'John'}) YIELD node RETURN node 

## Expected Behavior

Since the identProps argument is `ssid:'123'`, this should match to the just-created Jim node. The name property should be updated to 'John', as we would expect if running this Cypher equivalent:

```
MERGE (node:Person {ssid:'123'})
SET node.name = 'John'
```

## Actual Behavior (Mandatory)

The MERGE indeed matched to the existing Jim node (no new node was created), but the name property was not updated.  

## Other Notes

Looking at the PR for this, it looks like this behavior was intentional (see [testMergeNodeWithPreExisting()](https://github.com/neo4j-contrib/neo4j-apoc-procedures/pull/508/files#diff-0357b9c1505ecd881eaaa564ca7e2107R47), though there doesn't seem to be any documentation about this behavior, nor does it seem intuitive (this would behave like an ON CREATE SET).

### Versions

  - OS: Mac OS 10.12.6
  - Neo4j: 3.5.4 Enterprise
  - Neo4j-Apoc: 3.5.0.3

### PR DESCRIPTION
**Pull request by InverseFalcon**
Fixes #1182 
New procedures added, old procedures unchanged (some refactoring, but no changes to behavior)

These both accept a config map that can take 'onMatchProps' and 'onCreateProps' map properties.

These procs can also match to multiple existing nodes unlike apoc.merge.node() and apoc.merge.relationship().

Documentation to be checked in separately.

### CODE REVIEW

**Comment by jexp** [comment]
Are we good with this?

File: src/main/java/apoc/merge/Merge.java [hunk 1]
```diff
@@ -25,26 +25,49 @@
...
+    @Procedure(value="apoc.merge.node.eager", mode = Mode.WRITE, eager = true)
+    @Description("apoc.merge.node.eager(['Label'], identProps:{key:value, ...}, config:{onCreateProps:{key:value,...}, onMatchProps:{key:value,...}}) - merge node with dynamic labels, with support for setting properties ON CREATE or ON MATCH")
+    public Stream<NodeResult> nodeEager(@Name("label") List<String> labelNames, @Name("identProps") Map<String, Object> identProps, @Name(value="config", defaultValue = "{}") Map<String, Object> config) {
+        if ((identProps==null) || (identProps.isEmpty())) {
+            throw new IllegalArgumentException("you need to supply at least one identifying property for a merge");
+        }
+
+        validateKeys(config.keySet(), Lists.newArrayList("onCreateProps", "onMatchProps"));
+
+        String labels = labelNames.stream().map(s -> wrapInBacktics(s)).collect(Collectors.joining(":"));
```
Comment (jexp) [inline src/main/java/apoc/merge/Merge.java]: there is Util.quote which does a better job of this.

File: src/main/java/apoc/merge/Merge.java [hunk 2]
```diff
@@ -25,26 +25,49 @@
...
     }
 
     @Procedure(mode = Mode.WRITE)
-    @Description("apoc.merge.relationship(startNode, relType,  {key:value, ...}, {key:value, ...}, endNode) - merge relationship with dynamic type")
+    @Description("apoc.merge.relationship(startNode, relType,  identProps:{key:value, ...}, {key:value, ...}, endNode) - merge relationship with dynamic type")
     public Stream<RelationshipResult> relationship(@Name("startNode") Node startNode, @Name("relationshipType") String relType,
                                                    @Name("identProps") Map<String, Object> identProps, @Name("props") Map<String, Object> props, @Name("endNode") Node endNode) {
         String identPropsString = buildIdentPropsString(identProps);
 
         Map<String, Object> params = new HashMap<>();
```
Comment (jexp) [inline src/main/java/apoc/merge/Merge.java]: easier with Util.map(args)

File: src/main/java/apoc/merge/Merge.java [hunk 3]
```diff
@@ -53,18 +76,47 @@ private String wrapInBacktics(String s) {
         return Stream.of(new RelationshipResult(rel));
     }
 
+    @Procedure(value = "apoc.merge.relationship.eager", mode = Mode.WRITE, eager = true)
+    @Description("apoc.merge.relationship.eager(startNode, relType,  identProps:{key:value, ...}, config:{onCreateProps:{key:value, ...}, onMatchProps:{key:value, ...}}, endNode) - merge relationship with dynamic type, with support for setting properties ON CREATE or ON MATCH")
+    public Stream<RelationshipResult> relationshipEager(@Name("startNode") Node startNode, @Name("relationshipType") String relType,
+                                                   @Name("identProps") Map<String, Object> identProps, @Name("config") Map<String, Object> config, @Name("endNode") Node endNode) {
+        validateKeys(config.keySet(), Lists.newArrayList("onCreateProps", "onMatchProps"));
```
Comment (jexp) [inline src/main/java/apoc/merge/Merge.java]: shouldn't this just delegate to the non-eager version?

File: src/main/java/apoc/merge/Merge.java [hunk 4]
```diff
@@ -25,26 +25,49 @@
...
         Node node = Iterators.single(db.execute(cypher, params ).columnAs("n"));
         return Stream.of(new NodeResult(node));
     }
 
-    private String wrapInBacktics(String s) {
-        return "`" + s + "`";
+    @Procedure(value="apoc.merge.node.eager", mode = Mode.WRITE, eager = true)
+    @Description("apoc.merge.node.eager(['Label'], identProps:{key:value, ...}, config:{onCreateProps:{key:value,...}, onMatchProps:{key:value,...}}) - merge node with dynamic labels, with support for setting properties ON CREATE or ON MATCH")
+    public Stream<NodeResult> nodeEager(@Name("label") List<String> labelNames, @Name("identProps") Map<String, Object> identProps, @Name(value="config", defaultValue = "{}") Map<String, Object> config) {
+        if ((identProps==null) || (identProps.isEmpty())) {
```
Comment (jexp) [inline src/main/java/apoc/merge/Merge.java]: shouldn't this just delegate to the non-eager version?

File: src/main/java/apoc/merge/Merge.java (hunk 3 above)
Comment (InverseFalcon) [reply src/main/java/apoc/merge/Merge.java]: The non-eager version can only merge a single result (see `Iterators.single()` usage) and doesn't handle ON MATCH SET.

File: src/main/java/apoc/merge/Merge.java (hunk 4 above)
Comment (InverseFalcon) [reply src/main/java/apoc/merge/Merge.java]: The non-eager version can only merge a single result (see `Iterators.single()` usage) and doesn't handle ON MATCH SET.

File: src/main/java/apoc/merge/Merge.java (hunk 1 above)
Comment (InverseFalcon) [reply src/main/java/apoc/merge/Merge.java]: In place of the wrapInBacktics() call? Okay, will replace.

File: src/main/java/apoc/merge/Merge.java (hunk 2 above)
Comment (InverseFalcon) [reply src/main/java/apoc/merge/Merge.java]: For building out the `params` map? Got it.

File: src/main/java/apoc/merge/Merge.java (hunk 4 above)
Comment (jexp) [reply src/main/java/apoc/merge/Merge.java]: but then we should add both there

File: src/main/java/apoc/merge/Merge.java [hunk 5]
```diff
@@ -22,49 +23,88 @@
...
         params.put("endNode", endNode);
 
-        final String cypher = "WITH $startNode as startNode, $endNode as endNode MERGE (startNode)-[r:"+ wrapInBacktics(relType) +"{"+identPropsString+"}]->(endNode) ON CREATE SET r+= $props RETURN r";
+        final String cypher = "WITH $startNode as startNode, $endNode as endNode MERGE (startNode)-[r:"+ Util.quote(relType) +"{"+identPropsString+"}]->(endNode) ON CREATE SET r+= $props RETURN r";
         Relationship rel = Iterators.single(db.execute(cypher, params ).columnAs("r"));
         return Stream.of(new RelationshipResult(rel));
     }
 
+    @Procedure(value = "apoc.merge.relationship.eager", mode = Mode.WRITE, eager = true)
+    @Description("apoc.merge.relationship.eager(startNode, relType,  identProps:{key:value, ...}, config:{onCreateProps:{key:value, ...}, onMatchProps:{key:value, ...}}, endNode) - merge relationship with dynamic type, with support for setting properties ON CREATE or ON MATCH")
```
Comment (jexp) [inline src/main/java/apoc/merge/Merge.java]: I wouldn't put the extra props into config but into optional parameters which default to empty maps.

File: src/main/java/apoc/merge/Merge.java [hunk 6]
```diff
@@ -22,49 +23,88 @@
...
         final String cypher = "MERGE (n:" + labels + "{" + identPropsString + "}) ON CREATE SET n += $props RETURN n";
         Node node = Iterators.single(db.execute(cypher, params ).columnAs("n"));
         return Stream.of(new NodeResult(node));
     }
 
-    private String wrapInBacktics(String s) {
-        return "`" + s + "`";
+    @Procedure(value="apoc.merge.node.eager", mode = Mode.WRITE, eager = true)
+    @Description("apoc.merge.node.eager(['Label'], identProps:{key:value, ...}, config:{onCreateProps:{key:value,...}, onMatchProps:{key:value,...}}) - merge node with dynamic labels, with support for setting properties ON CREATE or ON MATCH")
+    public Stream<NodeResult> nodeEager(@Name("label") List<String> labelNames, @Name("identProps") Map<String, Object> identProps, @Name(value="config", defaultValue = "{}") Map<String, Object> config) {
```
Comment (jexp) [inline src/main/java/apoc/merge/Merge.java]: I wouldn't put the extra props into config but into optional parameters which default to empty maps.

File: src/main/java/apoc/merge/Merge.java [hunk 7]
```diff
@@ -22,49 +23,88 @@
             throw new IllegalArgumentException("you need to supply at least one identifying property for a merge");
         }
 
-        String labels = labelNames.stream().map(s -> wrapInBacktics(s)).collect(Collectors.joining(":"));
+        String labels = labelNames.stream().map(s -> Util.quote(s)).collect(Collectors.joining(":"));
```
Comment (jexp) [inline src/main/java/apoc/merge/Merge.java]: there is a method in Util for this.

**Comment by jexp** [comment]
Merged manually with some changes.
Removed the config maps.
Simplified the call to redirect.

**Pull request closed without merging**